### ISSUE TITLE
Unhandled exception in REST resource // Cannot deserialize instance of `java.lang.Long` out of START_OBJECT token

### ISSUE THREAD

**Issue by eadanilenko** [Graylog2/graylog2-server#9559]
Hi Team,

Have no idea how it happens and how to reproduce it, but from time to time I do able to see following exception:

```
2020-11-20T12:02:51.677+02:00 ERROR [AnyExceptionClassMapper] Unhandled exception in REST resource
java.lang.RuntimeException: IOException encountered while reading from a byte array input stream
        at org.mongojack.internal.stream.JacksonDBDecoder.decode(JacksonDBDecoder.java:67) ~[graylog.jar:?]
        at com.mongodb.DBDecoderAdapter.decode(DBDecoderAdapter.java:49) ~[graylog.jar:?]
        at com.mongodb.DBDecoderAdapter.decode(DBDecoderAdapter.java:29) ~[graylog.jar:?]
        at com.mongodb.operation.CommandResultArrayCodec.decode(CommandResultArrayCodec.java:52) ~[graylog.jar:?]
        at com.mongodb.operation.CommandResultDocumentCodec.readValue(CommandResultDocumentCodec.java:60) ~[graylog.jar:?]
        at org.bson.codecs.BsonDocumentCodec.decode(BsonDocumentCodec.java:84) ~[graylog.jar:?]
        at org.bson.codecs.BsonDocumentCodec.decode(BsonDocumentCodec.java:41) ~[graylog.jar:?]
        at org.bson.internal.LazyCodec.decode(LazyCodec.java:48) ~[graylog.jar:?]
        at org.bson.codecs.BsonDocumentCodec.readValue(BsonDocumentCodec.java:101) ~[graylog.jar:?]
        at com.mongodb.operation.CommandResultDocumentCodec.readValue(CommandResultDocumentCodec.java:63) ~[graylog.jar:?]
        at org.bson.codecs.BsonDocumentCodec.decode(BsonDocumentCodec.java:84) ~[graylog.jar:?]
        at org.bson.codecs.BsonDocumentCodec.decode(BsonDocumentCodec.java:41) ~[graylog.jar:?]
        at com.mongodb.internal.connection.ReplyMessage.<init>(ReplyMessage.java:51) ~[graylog.jar:?]
        at com.mongodb.internal.connection.InternalStreamConnection.getCommandResult(InternalStreamConnection.java:413) ~[graylog.jar:?]
        at com.mongodb.internal.connection.InternalStreamConnection.receiveCommandMessageResponse(InternalStreamConnection.java:309) ~[graylog.jar:?]
        at com.mongodb.internal.connection.InternalStreamConnection.sendAndReceive(InternalStreamConnection.java:259) ~[graylog.jar:?]
        at com.mongodb.internal.connection.UsageTrackingInternalConnection.sendAndReceive(UsageTrackingInternalConnection.java:99) ~[graylog.jar:?]
        at com.mongodb.internal.connection.DefaultConnectionPool$PooledConnection.sendAndReceive(DefaultConnectionPool.java:450) ~[graylog.jar:?]
        at com.mongodb.internal.connection.CommandProtocolImpl.execute(CommandProtocolImpl.java:72) ~[graylog.jar:?]
        at com.mongodb.internal.connection.DefaultServer$DefaultServerProtocolExecutor.execute(DefaultServer.java:226) ~[graylog.jar:?]
        at com.mongodb.internal.connection.DefaultServerConnection.executeProtocol(DefaultServerConnection.java:269) ~[graylog.jar:?]
        at com.mongodb.internal.connection.DefaultServerConnection.command(DefaultServerConnection.java:131) ~[graylog.jar:?]
        at com.mongodb.internal.connection.DefaultServerConnection.command(DefaultServerConnection.java:123) ~[graylog.jar:?]
        at com.mongodb.operation.QueryBatchCursor.getMore(QueryBatchCursor.java:260) ~[graylog.jar:?]
        at com.mongodb.operation.QueryBatchCursor.hasNext(QueryBatchCursor.java:138) ~[graylog.jar:?]
        at com.mongodb.client.internal.MongoBatchCursorAdapter.hasNext(MongoBatchCursorAdapter.java:54) ~[graylog.jar:?]
        at com.mongodb.DBCursor.hasNext(DBCursor.java:175) ~[graylog.jar:?]
        at org.mongojack.DBCursor.hasNext(DBCursor.java:330) ~[graylog.jar:?]
        at java.lang.Iterable.forEach(Iterable.java:74) ~[?:1.8.0_272]
        at org.graylog2.system.traffic.TrafficCounterService.clusterTrafficOfLastDays(TrafficCounterService.java:102) ~[graylog.jar:?]
        at org.graylog2.rest.resources.system.TrafficResource.get(TrafficResource.java:56) ~[graylog.jar:?]
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_272]
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_272]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_272]
        at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_272]
       at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory.lambda$static$0(ResourceMethodInvocationHandlerFactory.java:52) ~[graylog.jar:?]
        at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:124) ~[graylog.jar:?]
        at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:167) ~[graylog.jar:?]
        at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$TypeOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:219) ~[graylog.ja
r:?]
        at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:79) ~[graylog.jar:?]
        at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:469) ~[graylog.jar:?]
        at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:391) ~[graylog.jar:?]
        at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:80) ~[graylog.jar:?]
        at org.glassfish.jersey.server.ServerRuntime$1.run(ServerRuntime.java:253) [graylog.jar:?]
        at org.glassfish.jersey.internal.Errors$1.call(Errors.java:248) [graylog.jar:?]
        at org.glassfish.jersey.internal.Errors$1.call(Errors.java:244) [graylog.jar:?]
        at org.glassfish.jersey.internal.Errors.process(Errors.java:292) [graylog.jar:?]
        at org.glassfish.jersey.internal.Errors.process(Errors.java:274) [graylog.jar:?]
        at org.glassfish.jersey.internal.Errors.process(Errors.java:244) [graylog.jar:?]
        at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:265) [graylog.jar:?]
        at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:232) [graylog.jar:?]
        at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:680) [graylog.jar:?]
        at org.glassfish.jersey.grizzly2.httpserver.GrizzlyHttpContainer.service(GrizzlyHttpContainer.java:356) [graylog.jar:?]
        at org.glassfish.grizzly.http.server.HttpHandler$1.run(HttpHandler.java:200) [graylog.jar:?]
        at com.codahale.metrics.InstrumentedExecutorService$InstrumentedRunnable.run(InstrumentedExecutorService.java:180) [graylog.jar:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_272]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_272]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_272]
Caused by: com.fasterxml.jackson.databind.exc.MismatchedInputException: Cannot deserialize instance of `java.lang.Long` out of START_OBJECT token
 at [Source: de.undercouch.bson4jackson.io.LittleEndianInputStream@52ff6b69; pos: 55] (through reference chain: org.graylog2.system.traffic.AutoValue_TrafficDto$Builder["decoded"]->java.util.LinkedHashMap["graylog"])
        at com.fasterxml.jackson.databind.exc.MismatchedInputException.from(MismatchedInputException.java:63) ~[graylog.jar:?]
        at com.fasterxml.jackson.databind.DeserializationContext.reportInputMismatch(DeserializationContext.java:1343) ~[graylog.jar:?]
        at com.fasterxml.jackson.databind.DeserializationContext.handleUnexpectedToken(DeserializationContext.java:1139) ~[graylog.jar:?]
        at com.fasterxml.jackson.databind.DeserializationContext.handleUnexpectedToken(DeserializationContext.java:1093) ~[graylog.jar:?]
        at com.fasterxml.jackson.databind.deser.std.NumberDeserializers$LongDeserializer._parseLong(NumberDeserializers.java:593) ~[graylog.jar:?]
        at com.fasterxml.jackson.databind.deser.std.NumberDeserializers$LongDeserializer.deserialize(NumberDeserializers.java:557) ~[graylog.jar:?]
        at com.fasterxml.jackson.databind.deser.std.NumberDeserializers$LongDeserializer.deserialize(NumberDeserializers.java:535) ~[graylog.jar:?]
        at com.fasterxml.jackson.databind.deser.std.MapDeserializer._readAndBindStringKeyMap(MapDeserializer.java:527) ~[graylog.jar:?]
        at com.fasterxml.jackson.databind.deser.std.MapDeserializer.deserialize(MapDeserializer.java:364) ~[graylog.jar:?]
        at com.fasterxml.jackson.databind.deser.std.MapDeserializer.deserialize(MapDeserializer.java:29) ~[graylog.jar:?]
        at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeSetAndReturn(MethodProperty.java:158) ~[graylog.jar:?]
        at com.fasterxml.jackson.databind.deser.BuilderBasedDeserializer.vanillaDeserialize(BuilderBasedDeserializer.java:269) ~[graylog.jar:?]
        at com.fasterxml.jackson.databind.deser.BuilderBasedDeserializer.deserialize(BuilderBasedDeserializer.java:193) ~[graylog.jar:?]
        at com.fasterxml.jackson.databind.ObjectMapper._readValue(ObjectMapper.java:3985) ~[graylog.jar:?]
        at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:2343) ~[graylog.jar:?]
        at org.mongojack.internal.stream.JacksonDBDecoder.decode(JacksonDBDecoder.java:80) ~[graylog.jar:?]
        at org.mongojack.internal.stream.JacksonDBDecoder.decode(JacksonDBDecoder.java:64) ~[graylog.jar:?]
        ... 57 more
```

Maybe someone knows what is it and how to fix it?

Thanks!

* Graylog Version: 4.0.0-8
* Java Version: openjdk version 1.8.0_272
* Elasticsearch Version: 6.8.13
* MongoDB Version: 4.2
* Operating System: CentOS7
* Browser version: Google Chrome 86.0.4240.183

**Comment by patrickmann** [comment]
Based on the stack trace the scenario is:
- call `GET /system/cluster/traffic`
- which reads a number of documents from the `traffic` collection
- deserializing of a `decoded` field in one of the documents fails, due to an invalid numeric value

Here is an example of a valid `traffic` document:
```
{
    "_id" : ObjectId("611510880630020c06e39794"),
    "bucket" : ISODate("2021-08-12T12:00:00.000Z"),
    "decoded" : {
        "7e2d581d-ca24-46e9-b3b9-e9cc79eedf7e" : NumberLong(14507)
    },
    "input" : {
        "7e2d581d-ca24-46e9-b3b9-e9cc79eedf7e" : NumberLong(11409)
    },
    "output" : {
        "7e2d581d-ca24-46e9-b3b9-e9cc79eedf7e" : NumberLong(17387)
    }
}
```
We are just extracting the value from the DB as is. For deserialization to fail, the value in the DB or the value returned from the DB query must be corrupted. We shouldn't be able to serialize and write an invalid value in the first place, so where is it introduced?

**Comment by patrickmann** [comment]
This was also [reported](https://community.graylog.org/t/apierror-on-system-cluster-traffic/15531) in the community back in May '20.
Unfortunately, no solution was identified at the time.

**Comment by mpfz0r** [comment]
@eadanilenko If this happens again, it would be helpful to see the entire `traffic` collection, to get to the root of the problem.

e.g. running something like `mongo graylog --eval 'db.traffic.find().toArray()'`

**Comment by patrickmann** [comment]
> @eadanilenko If this happens again, it would be helpful to see the entire `traffic` collection, to get to the root of the problem.
> 
> e.g. running something like `mongo graylog --eval 'db.traffic.find().toArray()'`

I have also requested the data from user [HansWurst](https://community.graylog.org/u/hanswurst/summary) who reported this too.

**Comment by pdcemulator** [comment]
Hi Patrick,

this is a traffic document from my system:
```
{
"_id" : ObjectId("611b73880eae08a62c73b4c1"), 
"bucket" : ISODate("2021-08-17T08:00:00Z"), 
"decoded" : {
    "glhost2" : { "mydomain" : { "local" : NumberLong(7137662) } }, 
    "glhost1" : { "mydomain" : { "local" : NumberLong(6152202) } } }, 
"input" : { 
    "glhost2" : { "mydomain" : { "local" : NumberLong(6707357) } }, 
    "glhost1" : { "mydomain" : { "local" : NumberLong(5772190) } } }, 
"output" : { 
    "glhost2" : { "mydomain" : { "local" : NumberLong(8748067) } }, 
    "glhost1" : { "mydomain" : { "local" : NumberLong(7221993) } } } }
```

glhost1|2.mydomain.local is the FQDN of my two graylog cluster hosts. I also use the FQDN in `/etc/graylog/server/node-id`.
Then I removed the string `.mydomain.local` from node-id and restarted Graylog services.

Now the API calls are returning valid values and I have this document structure:

```
{ 
"_id" : ObjectId("611b77c00eae08a62c76e36d"), 
"bucket" : ISODate("2021-08-17T08:00:00Z"), 
"decoded" : { 
    "glhost2" : NumberLong(0), 
    "glhost1" : NumberLong(318438) }, 
"input" : { 
    "glhost2" : NumberLong(0), 
    "glhost1" : NumberLong(298675) }, 
"output" : { 
    "glhost2" : NumberLong(0), 
    "glhost1" : NumberLong(366663) } }
```

Is it a configuration error to use the FQDN as node-id or is it an error parsing that value?

Environment: Graylog 3.3.14+13e45b7 on glhost1.mydomain.local (Red Hat, Inc. 1.8.0_302 on Linux 4.18.0-326.el8.x86_64)

**Comment by patrickmann** [comment]
@pdcemulator There you have it. We obviously didn't account for node-ids containing dots. Since we allow it, we should deal with it appropriately. Stay tuned ...

**Comment by eadanilenko** [comment]
Sorry guys for delay in response, was on PTO with limited access.
I do have FQDN as a node-id in my environment as well. The workaround looks like works, there are no errors so far.

**Comment by patrickmann** [comment]
Fixed  in #11176 and should be available in the next release.
